File: backend/app/api/activity.py
"""Activity API - Unified activity feed endpoint.

Provides:
- GET /api/activity - Paginated activity feed from all sources
"""

from __future__ import annotations

import asyncio
from typing import Any, Literal

from fastapi import APIRouter, Query
from pydantic import BaseModel

from ..storage import activity as activity_store
from ..storage.activity.types import ActivityEvent as StorageActivityEvent

router = APIRouter()


# ============================================================================
# Request/Response Models
# ============================================================================


class ActivityEvent(BaseModel):
    """A single activity event."""

    type: Literal["task", "backup", "git"]
    message: str
    timestamp: str | None
    project_id: str
    metadata: dict[str, Any]


class ActivityFeedResponse(BaseModel):
    """Response for activity feed."""

    items: list[ActivityEvent]
    total: int
    limit: int
    offset: int
    has_more: bool


# ============================================================================
# Helpers
# ============================================================================

_VALID_TYPES = {"task", "backup", "git"}


def _parse_event_types(types: str | None) -> list[str] | None:
    """Parse and validate comma-separated event type filter string."""
    if not types:
        return None
    parsed = [t.strip() for t in types.split(",") if t.strip()]
    valid = [t for t in parsed if t in _VALID_TYPES]
    return valid or None


def _to_activity_event(e: StorageActivityEvent) -> ActivityEvent:
    """Convert a raw event dict to an ActivityEvent model."""
    return ActivityEvent(
        type=e["type"],
        message=e["message"],
        timestamp=e["timestamp"],
        project_id=e["project_id"],
        metadata=e["metadata"],
    )


# ============================================================================
# Endpoints
# ============================================================================


@router.get("/activity")
async def get_activity_feed(
    project_id: str | None = Query(default=None, description="Filter by project ID"),
    limit: int = Query(default=50, le=100, ge=1, description="Max events per page"),
    offset: int = Query(default=0, ge=0, description="Pagination offset"),
    types: str | None = Query(
        default=None,
        description="Comma-separated event types: task,backup,git",
    ),
) -> ActivityFeedResponse:
    """Get unified activity feed from all sources.

    Aggregates:
    - Task changes (created, updated, completed, blocked, cancelled)
    - Backup events (completed/failed backups)
    - Git commits (read from repository history)

    Results are sorted by timestamp (newest first) and paginated.
    """
    event_types = _parse_event_types(types)
    events, total = await asyncio.to_thread(
        activity_store.get_aggregated_activity,
        project_id=project_id,
        limit=limit,
        offset=offset,
        event_types=event_types,
    )
    return ActivityFeedResponse(
        items=[_to_activity_event(e) for e in events],
        total=total,
        limit=limit,
        offset=offset,
        has_more=offset + len(events) < total,
    )

File: frontend/lib/api/activity.ts
/**
 * Activity API - Unified activity feed for dashboard.
 */

import { buildQueryString, fetchWithErrorHandling } from './utils'

export type ActivityEventType = 'task' | 'backup' | 'git'

export interface ActivityEvent {
  type: ActivityEventType
  message: string
  timestamp: string | null
  project_id: string
  metadata: {
    task_id?: string
    backup_id?: string
    commit_sha?: string
    status?: string
    title?: string
    action?: string
    backup_type?: string
    size_bytes?: number
    repo_name?: string
    author_name?: string
    files_changed?: number
    insertions?: number
    deletions?: number
  }
}

export interface ActivityFeedResponse {
  items: ActivityEvent[]
  total: number
  limit: number
  offset: number
  has_more: boolean
}

export interface FetchActivityOptions {
  project_id?: string
  limit?: number
  offset?: number
  types?: ActivityEventType[]
}

/**
 * Fetch activity feed from the API.
 */
export async function fetchActivity(
  options: FetchActivityOptions = {},
): Promise<ActivityFeedResponse> {
  const query = buildQueryString({
    project_id: options.project_id,
    limit: options.limit ?? 50,
    offset: options.offset,
    types: options.types?.join(','),
  })

  return fetchWithErrorHandling<ActivityFeedResponse>(`/api/activity${query}`, {
    errorMessage: 'Failed to fetch activity feed',
  })
}
